GitCommands/Git/GitRevision.cs

```diff
@@ -86,6 +84,76 @@ public BuildInfo BuildStatus
         public string Subject { get; set; } = "";
         [CanBeNull]
         public string Body { get; set; }
+
+        [CanBeNull]
+        public string BodySummary
+        {
+            get
+            {
+                if (string.IsNullOrWhiteSpace(Body) || !HasMultiLineMessage)
+                {
+                    return Body;
+                }
+
+                return _bodySummary ?? (_bodySummary = BuildSummary(Body.TrimEnd()));
+            }
+        }
+
+        private static string BuildSummary(string body)
```

Comment:
In this method, if the `body` does not need to be changed (which I assume is the majority case) can you avoid the extra work and allocation by detecting this and just returning body?

Reply:
Avoiding the extra cost will be difficult (and I don't worth the effort because most of the time cpu is spent during display of the tooltip).

But avoiding the allocation could be done easily. And it's a good idea.

But I think we shouldn't be afraid to keep the value in memory because they are garbage collected when the revision grid is refreshed, I think.

Reply:
If you have 1,000 commits, you will on average only populate a handful of these summary properties. That's a lot of null values, taking up 8 bytes each on a 64-bit machine. Reducing the compactness of objects has knock on effects to cache efficiency and the like, when iterating through revisions. These objects should be as small as possible.

Reply:
Should there be a separate tooltip cache?

Reply:
I don't think there's any need to cache the tooltips. There are very few of them, they are quick to create, are created slowly (in response to user actions) and when you consider that they're just not large enough to worry about keeping around. The rest of the PR is good -- I just wouldn't worry about caching.

Reply:
Since nothing `GitRevision`-specific is done here anymore, this function should be moved to a new class.
We should not bloat this class with displaying stuff.
Then this function can be used to truncate all tooltips in `RevisionGridToolTipProvider` again.

Reply:
@mstv 👍

Reply:
I pushed a refactor in which I moved `GitRevision.BuildSummary(string)` method out to its own class.
I updated callsites, but tests need to be updated to reflect the new code structure.